### ISSUE TITLE
Docker container support for RTX 2070

### ISSUE THREAD

**Issue by NoelKennedy** [tensorflow/tensorflow#24045]
**System information**
- OS Platform and Distribution : I've tried two with the same result: Linux Ubuntu 16.04 & Linux Ubuntu 18.04
- TensorFlow installed from docker hub :  tensorflow/tensorflow:latest-gpu-py3
- TensorFlow version: 1.12
- Python version: 3
- Installed using docker : tensorflow/tensorflow:latest-gpu-py3
- CUDA/cuDNN version: 
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, **release 9.0, V9.0.176**

- GPU model and memory: dual RTX 2070 8gb

I'm trying to set up TF through a docker image and I'm having problems getting TF working with the GPUs.  I get a segfault shortly after TF allocates memory space on the GPUs.  I'm thinking the drivers are the issue at this point.

I would like to clarify the pre-requisites for the host machine to be capable of running the TF docker image.  The info on the [site](https://www.tensorflow.org/install/docker) seems pretty straight forward, but when you look at the detail, it becomes confusing (to me at least) and slightly contradictory.  There appears to be a discrepancy between what is stated in instructions, and where you end up if you follow the links in the instructions.  

The [pre-requisites](https://www.tensorflow.org/install/docker) for the docker image state _only the NVIDIA® GPU driver is required on the host machine (the NVIDIA® CUDA® Toolkit does not need to be installed)_.

However, if you follow the links, you will end up on a page which instructions you to install the CUDA toolkit and CUDA, by implication making these a pre-requisite for using the TR docker container.

The hyperlink for 'NVIDIA® GPU driver' links to an [NVIDIA FAQ](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#how-do-i-install-the-nvidia-driver) which recommends you use your package manager to install the NVidia drivers (which is what I have done, using the PPA repository ([Instructions here](http://www.linuxandubuntu.com/home/how-to-install-latest-nvidia-drivers-in-linux)).  However, if you click the link behind _package manager_, you end up [here](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#package-manager-installation) 
 an NVidia site which details how to install CUDA, not the NVidia graphics drivers! Also, if you check the pre-requisites of the CUDA installation guide you will see you need to install the CUDA toolkit, which seems to contradict the statement at the start:

_only the NVIDIA® GPU driver is required on the host machine (the NVIDIA® CUDA® Toolkit does not need to be installed)_

So do we need to install CUDA toolkit and CUDA and graphics drivers on the host?

Confusingly, there also seems to be something called a 'CUDA driver' which I assume is something different from the NVidia driver.  The [NVidia instructions ](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#how-do-i-install-the-nvidia-driver) are about installing the 'CUDA driver' on the host machine, not the NVidia driver??

_The recommended way is to use your package manager and install the cuda-drivers package (or equivalent)_

Please can you advise how I should set up the host machine to make it work with the TF docker images?

Thanks!

**Comment by wdirons** [comment]
CUDA driver and nvidia driver are the same thing, just some people call it different things.

What you need on the host is the nvidia driver. The CUDA toolkit belongs in the docker container. 

1. Install the nvidia driver
2. Reboot
3. Verify with `nvidia-smi` the driver is working on the host
4. Install docker
5. Install nvidia-docker
6. Verify with `docker run --runtime=nvidia --rm nvidia/cuda:9.0-base nvidia-smi` that nvidia is working inside of the container.

If all this works, you should be able to run tensorflow inside of a container.

**Comment by NoelKennedy** [comment]
I've discovered my setup appears to work with the TF container version 1.10, but not with the later containers: TF 1.11 or 1.12.0.

I tested with nvidia-smi and running ./deviceQuery and ./bandwidthTest from the CUDA examples on the host & from inside the container.  I've found that even if these commands work, TF will still segfault.  I test TF by running the Jupyter notebook that comes with the image: 3_mnist_from_scratch.ipynb .  This segfaults on this command (16th cell)

# Run the graph and fetch some of the nodes.
_, l, lr, predictions = s.run(
  [optimizer, loss, learning_rate, train_prediction],
  feed_dict=feed_dict)

I've tried the follow drivers (I'm using dual RTX 2070 cards)

Driver source [PPA](https://launchpad.net/~graphics-drivers/+archive/ubuntu/ppa):
396 - Don't recognise my RTX 2070 (this was expected, wrong architecture)
410 - At time of writing, this installs driver version 410.78. nvidia-smi, and cuda samples succeed in both host and container.  Segfaults in container TF 1.12.0, 1.11.0.  Succeeds in container TF 1.10.0
415 - nvidia-smi, and cuda samples work in host and container.  Segfaults on TF 1.12.0.  Not tested any other TF version.

I did install CUDA 9.2 on the host, but I don't know if that helped or made no difference.

**Comment by wdirons** [comment]
Do you have the stack trace of the segfault? That may help track down if there is already an issue opened for the segfault problem. (Since you got docker working, you might want to re-title the issue if your allowed to.)

**Comment by wdirons** [comment]
The issue might be because the RTX card requires the CUDA 10 toolkit , and the TensorFlow Dockerfile includes CUDA 9.0 at this time. You might have to build TensorFlow from source.

See these issues:
https://github.com/tensorflow/tensorflow/issues/22900
https://github.com/tensorflow/tensorflow/issues/22706

Although I can't explain why 1.10 would have worked and later versions would not have.

**Comment by NoelKennedy** [comment]
Thanks William!  

The fact that TF is expected CUDA 9 and the RTX 2070 wants CUDA 10 explains a lot.  Like you say the only weird thing is how TF 1.10 works at all.

I'm happy to wait for TF to support CUDA 10 as it looks like it's going to happen soonish #22706
